src/utils/reductions.hpp

```diff
@@ -59,20 +59,26 @@ struct ReductionBase {
   T val;
 #ifdef MPI_PARALLEL
   MPI_Request req;
-  MPI_Comm comm;
+  std::shared_ptr<MPI_Comm> comm;
```

Comment:
Since this is a pointer now, it might also make sense to rename it `pcomm` for clarity.